### PR TITLE
Add AutoArchiveLogs option to send all workflow pod logs to Artifact Repository

### PR DESCRIPTION
**Pull request by irlevesque**
Fixes: https://github.com/argoproj/argo/issues/1790

### CODE REVIEW

File: workflow/config/config.go [hunk 1]
```diff
@@ -83,6 +83,8 @@ type KubeConfig struct {
 type ArtifactRepository struct {
 	// ArchiveLogs enables log archiving
 	ArchiveLogs *bool `json:"archiveLogs,omitempty"`
+	// AutoArchiveLogs enables log archiving for all tasks regardless of workflow configuration
+	AutoArchiveLogs *bool `json:"autoArchiveLogs,omitempty"`
```
Comment (simster7) [inline workflow/config/config.go]: Maybe `ForceArchiveLogs` is more explicit?

File: workflow/controller/workflowpod.go [hunk 2]
```diff
@@ -834,6 +834,9 @@ func (woc *wfOperationCtx) addArchiveLocation(pod *apiv1.Pod, tmpl *wfv1.Templat
 			break
 		}
 	}
+	if woc.artifactRepository.AutoArchiveLogs != nil && *woc.artifactRepository.AutoArchiveLogs {
+		needLocation = true
+	}
```
Comment (simster7) [inline workflow/controller/workflowpod.go]: Hmmm, currently it's not clear that even when enabling `AutoArchiveLogs` the user will also need to enable `ArchiveLogs` for the logs to be archived. Maybe a couple of lines below should actually be:

```go
	tmpl.ArchiveLocation = &wfv1.ArtifactLocation{
		ArchiveLogs: woc.artifactRepository.ArchiveLogs || woc.artifactRepository.AutoArchiveLogs,
	}
```

File: workflow/config/config.go (hunk 1 above)
Comment (irlevesque) [reply workflow/config/config.go]: How about `DefaultArchiveLogs`? This change wouldn't _force_ logs to be archived (users explicitly setting s3/artifactory destinations in template wouldn't be covered by this).

File: workflow/controller/workflowpod.go (hunk 2 above)
Comment (irlevesque) [reply workflow/controller/workflowpod.go]: Agreed on clarity. I think your proposal sounds reasonable.

**Comment by simster7** [comment]
On more thought, I'm unsure if a config in `ArtifactRepository` should be allowed to change the behavior of Workflows downstream (as opposed to only changing its own behavior). It seems to me that the Repository shouldn't be able to "force" Workflows that use it to do anything. In the case of `ArchiveLogs` the Repository simply "allows" log archiving – which is only affecting its own behavior.

What do you think? Can you think of another way to do this that would make the Workflow the active player in this?

I want to know what @jessesuen thinks of this too

**Comment by alexec** [comment]
I've just managed to re-implement this almost exactly the same in #1860

**Comment by alexec** [comment]
@irlevesque do you want to compare the two solutions, they're not exactly the same, as I re-use the `ArchiveLogs` flag, as this already work if there are any outputs.

**Comment by irlevesque** [comment]
Hi @alexec, my rationale for using a new variable was to avoid a change of behavior for existing users of the `ArchiveLogs` feature, who may have vested interest in maintaining the current functionality (and may indeed not want this new functionality).

**Pull request closed without merging**